### PR TITLE
36 configure build system and dependencies using nix

### LINKED ISSUE #36: Configure build system and dependencies using Nix
# Overview

Vmd is a component of the [Ghaf-system](https://github.com/tiiuae/ghaf) which is built and configured using the [Nix](https://nixos.wiki/) package manager or operating system. We need to be able to build vmd as part of the Ghaf using the same Nix-system. This requires defining a `flake.nix` file which defines dependencies and interfaces with build scripts for the project.

# Breakdown

- [x] #41
- [x] Define `flake.nix` in the root of the repository which implements building `vmd`.

# Deliverables

- A `flake.nix` file at the root of the repository.
- Build (pure, reproducable) on ghaf after cloning the vmd-repository.

### LINKED ISSUE #41: Resolve issues building against openSSL on NixOS.
## Overview

Simple `cargo build` fails on NixOS because of non-standard installation paths leading to incorrect configuration of environment variables.

We don't actually use openssl for anything, but the dependency is automatically added in the generated API bindings.

When I tried this with a `flake.nix`, `shell.nix` or `default.nix`  all of those were giving very cryptic errors so I figured I'd try to do a simple local build first.

Both `openssl` and `pkg-config` are installed on the system.

```
# /etc/nixos/configuration.nix
environment.systemPackages = with pkgs; [
    neovim # Do not forget to add an editor to edit configuration.nix! The Nano editor is also installed by default.
    wget
    curl
    openapi-generator-cli
    pkg-config
    openssl
    openssl.dev
  ];
```

## Reproduce

On NixOS:

```
git clone https://github.com/tiiuae/vmd/tree/36-configure-build-system-and-dependencies-using-nix
cd vmd
nix-build
```

## Trace

```
error: failed to run custom build command for `openssl-sys v0.9.87`

Caused by:
  process didn't exit successfully: `/home/juliuskoskela/Repos/vmd/target/debug/build/openssl-sys-3de660ecab229b6b/build-script-main` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR
  X86_64_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
  OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR
  X86_64_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
  OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_DIR
  X86_64_UNKNOWN_LINUX_GNU_OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_DIR
  OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  run pkg_config fail: `PKG_CONFIG_ALLOW_SYSTEM_CFLAGS="1" "pkg-config" "--libs" "--cflags" "openssl"` did not exit successfully: exit status: 1
  error: could not find system library 'openssl' required by the 'openssl-sys' crate

  --- stderr
  Package openssl was not found in the pkg-config search path.
  Perhaps you should add the directory containing `openssl.pc'
  to the PKG_CONFIG_PATH environment variable
  No package 'openssl' found


  --- stderr
  thread 'main' panicked at '

  Could not find directory of OpenSSL installation, and this `-sys` crate cannot
  proceed without this knowledge. If OpenSSL is installed and this crate had
  trouble finding it,  you can set the `OPENSSL_DIR` environment variable for the
  compilation process.

  Make sure you also have the development packages of openssl installed.
  For example, `libssl-dev` on Ubuntu or `openssl-devel` on Fedora.

  If you're in a situation where you think the directory *should* be found
  automatically, please open a bug at https://github.com/sfackler/rust-openssl
  and include information about your system as well as this message.

  $HOST = x86_64-unknown-linux-gnu
  $TARGET = x86_64-unknown-linux-gnu
  openssl-sys = 0.9.87

  ', /home/juliuskoskela/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-sys-0.9.87/build/find_normal.rs:190:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

## References

https://discourse.nixos.org/t/openssl-dependency-for-rust/3186
https://github.com/NixOS/nixpkgs/issues/54489
https://discourse.nixos.org/t/rust-openssl-woes/12340
https://discourse.nixos.org/t/openssl-dependency-for-rust/3186/4

### PR DESCRIPTION
**Pull request by juliuskoskela**
## Overview

Issue: https://github.com/tiiuae/vmd/issues/36

- [x] Create a `flake.nix` so that project can be built using `nix build` (mind the dash).

### CODE REVIEW

**Comment by juliuskoskela** [comment]
A note about the generated bindings. Apparently a `flake.nix` uses Git to see what source files should be included in the project. Thus generated API bindings have to be committed to the repository.

**Review by vadika — APPROVED** [review]
Looks good to me, except one thing -- may be it's good idea to squash five commits to one.

**Comment by juliuskoskela** [comment]
> Looks good to me, except one thing -- may be it's good idea to squash five commits to one.

I will squash the whole PR once it's approved.

**Comment by juliuskoskela** [comment]
Seems that `nix shell` command doesn't work. It doesn't give any error, but also doesn't load the developer shell.

**Comment by juliuskoskela** [comment]
> Seems that `nix shell` command doesn't work. It doesn't give any error, but also doesn't load the developer shell.

I misunderstood how the `nix shell` command should work. I guess it doesn't open the shell, just takes in a command to execute in it.

**Comment by juliuskoskela** [comment]
Generated API bindings are included in source control now. As far as I understand, nix requires the sources to be in Git's source control or at least I couldn't get the project to build otherwise.

File: flake.lock [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        "rev": "cfacdce06f30d2b68473a46042957675eebb3401",
+        "type": "github"
+      },
+      "original": {
+        "owner": "numtide",
+        "repo": "flake-utils",
+        "type": "github"
+      }
+    },
+    "flake-utils_2": {
```
Comment (mikatammi) [inline flake.lock]: Please get rid of this _2

File: flake.lock [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        "rev": "da67096a3b9bf56a91d16901293e51ba5b49a27e",
+        "type": "github"
+      },
+      "original": {
+        "owner": "nix-systems",
+        "repo": "default",
+        "type": "github"
+      }
+    },
+    "systems_2": {
```
Comment (mikatammi) [inline flake.lock]: and this

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,67 @@
+{
+  description = "Virtual Machine Daemon";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+
+    crane = {
+		url = "github:ipetkov/crane";
+		inputs.nixpkgs.follows = "nixpkgs";
```
Comment (mikatammi) [inline flake.nix]: You can get rid of the _2's above by adding here `inputs.flake-utils.follows = "flake-utils";`

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+					cargo
+				];
+
+				PKG_CONFIG_PATH = "${pkgs.openssl.dev}/lib/pkgconfig";
+			};
+		in {
+			checks = {
+				inherit package;
+			};
+
```
Comment (mikatammi) [inline flake.nix:50]: A hint, you could add formatter here to automatically prettify your nix files in this project, just like in ghaf:

https://github.com/tiiuae/ghaf/blob/main/flake.nix#L60

Then you could run `nix fmt *` to format your files

**Review by mikatammi — CHANGES_REQUESTED** [review]
I think the flake in general could be made much simpler. I was also thinking do we need or want to use crane to build the actual final package, however I can see the usefulness of using it for the development purposes

This is not all, I just briefly checked some of the Nix changes first. Otherwise a good start

I'd like to have this pattern which matches the nixpkgs repo, where the vmd has a separate nix-file (eg. vmd.nix), which could then be called through the callPackage-function. That ensures the overlays will get used when integrating this into ghaf.

**Comment by mikatammi** [comment]
> Generated API bindings are included in source control now. As far as I understand, nix requires the sources to be in Git's source control or at least I couldn't get the project to build otherwise.

You could also generate the bindings inside the derivation. We can have a deeper look at this

File: flake.nix (hunk 3 above)
Comment (juliuskoskela) [reply flake.nix]: Resolved in https://github.com/tiiuae/vmd/pull/42/commits/94f68ae6ca8b9552aada4402c3cf285e55499fa2

File: flake.nix (hunk 4 above)
Comment (juliuskoskela) [reply flake.nix:50]: Resolved in https://github.com/tiiuae/vmd/pull/42/commits/e309c277f1aa5a300be515c881c88eb5844eab28

**Comment by juliuskoskela** [comment]
> I think the flake in general could be made much simpler. I was also thinking do we need or want to use crane to build the actual final package, however I can see the usefulness of using it for the development purposes
> 
> This is not all, I just briefly checked some of the Nix changes first. Otherwise a good start
> 
> I'd like to have this pattern which matches the nixpkgs repo, where the vmd has a separate nix-file (eg. vmd.nix), which could then be called through the callPackage-function. That ensures the overlays will get used when integrating this into ghaf.

One thing to note is that `vmd` is ultimately 3 different programs. There will be the server, client-cli and client-gui.

_Maybe this is an argument to separate vmd-server and vmd-client(s) into separate repositories as originally planned? It seems they are very separate concerns from the perspective of building the Ghaf-system._

The way I see it is that the clients are obviously just normal packages that would be installed on any normal system through a package manager. The server on the other hand is different. It's a (system?) service which would be configured and enabled in the configuration on the host machine. It will run in `admin-vm` or something similar and the traffic will be routed from `netvm` (the single point of outside network traffic).

If the routing is going to happen just with local network as in `<IP>:<PORT> -> netvm -> localhost:<PORT> -> admin-vm(vmd)` (this might happen with shared memory or some other mechanism, as far as I know the design on inter-vm communication isn't finished yet), the configuration regarding `vmd` would span several modules (such as `netvm`) in the Ghaf system. Furthermore it would seem  that `vmd` should be defines as a [nixOS module](https://scvalex.net/posts/58/). Probably configuration of `vmd` would be part of the configuration of `admin-vm` if that will be the context we settle on (or are they actually the same thing?).

_Are flakes an modules somehow mutually exclusive and/or overlapping concepts?_

Here's a rough sketch on how a part of the configuration could look like:

```nix
{
  netvm = { config, pkgs, ... }: {
    routing.adminvm = {
      inbound-port = 1234;
      outbound-port = 6666; # localhost:6666
    };
    #...
  };
  
  adminvm = { config, pkgs, ... }: {
    services.vmd = {
      enable = true;
      package = pkgs.vmd;
      port = 6666;
    };
  };
  
  #...
}
```

In summary:

- We have two packages `vmd-client-cli` and `vmd-client-gui` which would be distributed as normal package derivations and are intended to be used on a "client" machine.
- We have a [module](https://nixos.wiki/wiki/NixOS_modules) which defines `vmd-server`. This service would be enabled and configured as a part of the host configuration of Ghaf.

**Comment by vilvo** [comment]
> 
> One thing to note is that `vmd` is ultimately 3 different programs. There will be the server, client-cli and client-gui.
> 
> _Maybe this is an argument to separate vmd-server and vmd-client(s) into separate repositories as originally planned? It seems they are very separate concerns from the perspective of building the Ghaf-system._

I think this is a bit misrepresentation of dependency between the `vmd-server` and `client-*`. `vmd-server` as such with no `client-*` is practically of no use. One could argue that `vmd-server` can have scripted tests but in essence those would just duplicate the `client-cli` which should also work as reference implementation on `vmd-server' client use. Given that, separating repositories sounds not not only bikeshedding but also ignores the depedency between the server and the client. Sure, design removes tight coupling between the server and the client but the dependency is there. 
 
At early stages of development where we can see also the generated bindings changing with implementation and further added functionality, it's clearer to change both server and the client at the same time. Option is separate repository and disciplined versioning to meet the dependency requirement management but I do not see the point.

This does not rule out separation of server and client(s) to separate repositories in the future - in fact, I've proposed we can do with git subtree split. Still, I do not support it *right now* as the packaging can be managing with respective nix-files and flake targets as was proposed.
 
> The way I see it is that the clients are obviously just normal packages that would be installed on any normal system through a package manager. The server on the other hand is different. It's a (system?) service which would be configured and enabled in the configuration on the host machine. It will run in `admin-vm` or something similar and the traffic will be routed from `netvm` (the single point of outside network traffic).

Yes and no. client-cli should go hand-in-hand with server integration in debug-builds but it could also be installed on other VMs to test routing...

> If the routing is going to happen just with local network as in `<IP>:<PORT> -> netvm -> localhost:<PORT> -> admin-vm(vmd)` (this might happen with shared memory or some other mechanism, as far as I know the design on inter-vm communication isn't finished yet), the configuration regarding `vmd` would span several modules (such as `netvm`) in the Ghaf system. Furthermore it would seem that `vmd` should be defines as a [nixOS module](https://scvalex.net/posts/58/). Probably configuration of `vmd` would be part of the configuration of `admin-vm` if that will be the context we settle on (or are they actually the same thing?).

... which means that client-cli should be possible to install as a package but also should be server dependency for now.
 
If we absolutely insist on separate repositories, I'd like to hear solid arguments on how it handles and solves problems at this point.

**Comment by juliuskoskela** [comment]
> If we absolutely insist on separate repositories, I'd like to hear solid arguments on how it handles and solves problems at this point.

I don't think separate repositories are a must at this point.

> ... which means that client-cli should be possible to install as a package but also should be server dependency for now.

Aren't both kind of possible by default if the client(s) are just packages?

**Comment by juliuskoskela** [comment]
> > Generated API bindings are included in source control now. As far as I understand, nix requires the sources to be in Git's source control or at least I couldn't get the project to build otherwise.
> 
> You could also generate the bindings inside the derivation. We can have a deeper look at this

I'd like to figure this one out before merging.

**Comment by vilvo** [comment]
> > ... which means that client-cli should be possible to install as a package but also should be server dependency for now.
> 
> Aren't both kind of possible by default if the client(s) are just packages?

yes

**Comment by juliuskoskela** [comment]
Resolved in https://github.com/tiiuae/vmd/pull/40
